src/widget/wtrackmenu.cpp

```diff
@@ -1627,6 +1657,181 @@ void WTrackMenu::slotClearAllMetadata() {
             &trackOperator);
 }
 
+namespace {
+
+class RemoveTrackFilesFromDiskTrackPointerOperation : public mixxx::TrackPointerOperation {
+  public:
+    QList<TrackRef> getTracksToPurge() const {
```

Comment:
Return by const-ref, even if implicitly shared.

Reply:
Are you sure? I'm confused why one would return a value by const reference. Wouldn't that reference have a high chance of becoming dangling in the future?

Reply:
No, common practice for getters. Return by value is only used for cheap copies and move (RVO).

Of course, the caller is responsible for taking lifetimes into account. Not checked as in Rust.

Reply:
> No, common practice for getters. Return by value is only used for cheap copies and move (RVO).

So should we always return by const ref if we return an existing list/vector/map (i.e. not constructed on-demand)? I  vaguely remember someone requested to change the const ref into a by-value return value on one of my PR some time ago, and that confused me but I assumed it was due to some c++ weirdness that I didn't know about.

Reply:
Am I getting it right: `const QList<TrackRef>& getTracksToPurge() const {`?

Reply:
References to members are returned by const-ref, local variables by value.

Exception: Members of objects that are used by multiple threads where the reference would otherwise outlive the lock guard! And even if you don't need a guard (smart pointers & implicitly shared object) you should return by value for a taking a snapshot of the state.

If you are unsure about lifetimes then please stay away from C/C++ and better use Rust ;)

Reply:
Thanks for the explanation. I was not aware that this was a getter function that just returns a member (my fault for not actually looking at the code), returning by const-ref makes sense then.

> If you are unsure about lifetimes then please stay away from C/C++ and better use Rust ;)

True, but that doesn't mean I don't want to learn. I'd say I'm pretty comfortable with lifetimes (because I learned rust) but this tripped me up so I figured I'd ask so I can improve my understanding. Thanks for the explanation again.

Reply:
@Swiftb0y You hopefully noticed that my sarcastic comment was not directed at you, Niko!

In C/C++ you have to do the work of the borrow checker in your mind, thereby reducing your mental capacity. Or you simply ignore it and your code will work by chance or 99.9999999% of the time. Managed, imperative languages have their own pitfalls.